### PR TITLE
feat: add ThisObjectValue

### PR DESCRIPTION
**Pull request by summer-ji-eng**
fullfill `ThisObjectValue`
- Refer instance variable
- Invoke class’s method with or without arguments
- Return the referenced class

### CODE REVIEW

File: src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  public abstract static class Builder {
+    public abstract Builder setType(TypeNode type);
+
+    public abstract ThisObjectValue autoBuild();
+
+    public ThisObjectValue build() {
+      ThisObjectValue thisObjectValue = autoBuild();
+      Preconditions.checkState(
+          TypeNode.isReferenceType(thisObjectValue.type()), "this can only refer to object types");
+      Preconditions.checkState(
```
Comment (summer-ji-eng) [inline src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:49]: I add this new check here to make sure the `this` type is belongs to custom class object. I see all the case of using this is inside of class definition, like `public Class Student {}`. This check is to exclude java build-in class like `java.lang.*`. [example](https://github.com/googleapis/gapic-generator-java/pull/137/files#diff-eb673d6d6c1d0400da66863c9b86eddfR39) [implementation](https://github.com/googleapis/gapic-generator-java/pull/137/files#diff-3eb77b740c6a645ccd1ef7aa836d7348R108)

But I am not pretty sure about this. Open to your advice and suggestion @miraleung @xiaozhenliu-gg5. Thanks a lot.

Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:50]: I think we don't need this check because `this` can be misused in many non-`java.lang` classes, which span all Java libs or other deps. Since checking the entire space is not feasible, we may as well not have this check.

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    ClassDefinition classDefinition =
+        ClassDefinition.builder()
+            .setPackageString("com.google.example.library.v1.stub")
+            .setName("LibraryServiceStub")
+            .setScope(ScopeNode.PUBLIC)
+            .build();
+    ConcreteReference ref = ConcreteReference.withClazz(classDefinition.getClass());
+    TypeNode typeNode = TypeNode.withReference(ref);
+    ThisObjectValue thisObjectValue = ThisObjectValue.withType(TypeNode.withReference(ref));
+    Assert.assertEquals(thisObjectValue.value(), "this");
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:48]: Nit: Since we hardcoded `"this"` as the return value for `value()`, we probably don't need to test this condition.

File: src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  public static ThisObjectValue withType(TypeNode type) {
+    return builder().setType(type).build();
+  }
+
+  private static Builder builder() {
+    return new AutoValue_ThisObjectValue.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:41]: Let's remove `public` to make this "private."

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+package com.google.api.generator.engine.ast;
+
+import static org.junit.Assert.assertThrows;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ThisObjectValueTest {
+  @Test
+  public void testThisObjectValueTest() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:24]: Nit: Let's follow the existing `validFoobar_reasonOrEdgeCase` convention. In g3, folks are migrating away from having "test" as a suffix, since this is already a test class.

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            .build();
+    TypeNode typeNode = TypeNode.withReference(ref);
+    ThisObjectValue thisObjectValue = ThisObjectValue.withType(TypeNode.withReference(ref));
+
+    Assert.assertEquals(thisObjectValue.value(), "this");
+    Assert.assertEquals(thisObjectValue.type(), typeNode);
+  }
+
+  @Test
+  public void testThisObjectValueTest_invalid() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:38]: The test case has the name "invalid" but the behavior looks like it succeeds?

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            .build();
+    ConcreteReference ref = ConcreteReference.withClazz(classDefinition.getClass());
+    TypeNode typeNode = TypeNode.withReference(ref);
+    ThisObjectValue thisObjectValue = ThisObjectValue.withType(TypeNode.withReference(ref));
+    Assert.assertEquals(thisObjectValue.value(), "this");
+    Assert.assertEquals(thisObjectValue.type(), typeNode);
+  }
+
+  @Test
+  public void testThisObjectValueTest_invalid_type() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:53]: Can we make this name more specific, e.g. `invalidThisObjectValue_primitiveType`?

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        () -> {
+          ThisObjectValue.withType(TypeNode.withReference(ref));
+        });
+    assertThrows(
+        IllegalStateException.class,
+        () -> {
+          ThisObjectValue.withType(TypeNode.BOOLEAN);
+        });
+  }
+}
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:66]: Add a test for `this` on a null type?

File: src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java [hunk 8]
```diff
@@ -1654,6 +1655,56 @@ public void writeClassDefinition_statementsAndMethods() {
             "}\n"));
   }
 
+  @Test
+  public void writeMethodReturnThisObjectValue() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java:1659]: Nit: The name of this test is valid, but since we're aiming to test `this`, let's call it something like `writeThisObjectValue_methodReturn`. Same below.

File: src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java [hunk 9]
```diff
@@ -1654,6 +1655,56 @@ public void writeClassDefinition_statementsAndMethods() {
...
+            .build();
+    methodDefinition.accept(writerVisitor);
+    assertEquals(
+        writerVisitor.write(),
+        String.format("public Student apply() {\n" + "return this;\n" + "}\n"));
+  }
+
+  @Test
+  public void writeExprStatementForThisObjectValue() {
+    // Example of ThisObjectValue access its fields
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java:1679]: Nit: Usual grammar rules apply to comments, e.g. Example of ThisObjectValue accessing its fields. <-
Same below please.

File: src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java (hunk 1 above)
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:50]: Sounds reasonable. removed. Thanks.

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java (hunk 4 above)
Comment (summer-ji-eng) [reply src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:24]: Gotcha. I don't have change to make changes on g3 yet :( 👍

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java (hunk 2 above)
Comment (summer-ji-eng) [reply src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:48]: I will remove the check to make consistent with other tests.

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java (hunk 7 above)
Comment (summer-ji-eng) [reply src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:66]: agree

File: src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class ThisObjectValue implements ObjectValue {
+  private static final String THIS_VALUE = "this";
+
+  public abstract TypeNode type();
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:24]: `@Override` for `ObjectValue`.

File: src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            .setName("Student")
+            .setPakkage("com.google.example.examples.v1")
+            .build();
+    TypeNode typeNode = TypeNode.withReference(ref);
+    ThisObjectValue.withType(typeNode);
+    // No exception thrown, we're good.
+  }
+
+  @Test
+  public void invalidThisObjectValue_NonReferenceType() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/ThisObjectValueTest.java:35]: Nit: "_nonReferenceType", same with _nullType below.

File: src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setType(TypeNode type);
+
+    abstract ThisObjectValue autoBuild();
+
+    private ThisObjectValue build() {
+      ThisObjectValue thisObjectValue = autoBuild();
+      Preconditions.checkState(
+          TypeNode.isReferenceType(thisObjectValue.type()), "this can only refer to object types");
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java:49]: Nit: Rephrase to make it clear that "this" is `this`, e.g. `The "this" object can only ...`
